File: NickvisionMoney.Shared/Controllers/MainWindowController.cs
using Nickvision.Aura;
using Nickvision.Aura.Events;
using NickvisionMoney.Shared.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using static Nickvision.Aura.Localization.Gettext;

namespace NickvisionMoney.Shared.Controllers;

/// <summary>
/// A controller for a MainWindow
/// </summary>
public class MainWindowController : IDisposable
{
    private bool _disposed;
    private string? _fileToLaunch;
    private List<AccountViewController> _openAccounts;

    /// <summary>
    /// A function for getting a password for an account
    /// </summary>
    public Func<string, Task<string?>>? AccountLoginAsync { get; set; }

    /// <summary>
    /// Gets the AppInfo object
    /// </summary>
    public AppInfo AppInfo => Aura.Active.AppInfo;
    /// <summary>
    /// The preferred theme of the application
    /// </summary>
    public Theme Theme => Configuration.Current.Theme;
    /// <summary>
    /// Whether to use native digits
    /// </summary>
    public bool UseNativeDigits => Configuration.Current.UseNativeDigits;
    /// <summary>
    /// The list of recent accounts
    /// </summary>
    public List<RecentAccount> RecentAccounts => Configuration.Current.RecentAccounts;
    /// <summary>
    /// The number of open accounts
    /// </summary>
    public int NumberOfOpenAccounts => _openAccounts.Count;

    /// <summary>
    /// Occurs when a notification is sent
    /// </summary>
    public event EventHandler<NotificationSentEventArgs>? NotificationSent;
    /// <summary>
    /// Occurs when an account is added
    /// </summary>
    public event EventHandler<EventArgs>? AccountAdded;
    /// <summary>
    /// Occurs when the recent accounts list is changed
    /// </summary>
    public event EventHandler<EventArgs>? RecentAccountsChanged;

    /// <summary>
    /// Constructs a MainWindowController
    /// </summary>
    /// <param name="args">Command-line arguments</param>
    public MainWindowController(string[] args)
    {
        _disposed = false;
        if (args.Length > 0)
        {
            _fileToLaunch = (Path.Exists(args[0]) && Path.GetExtension(args[0]).ToLower() == ".nmoney") ? args[0] : null;
        }
        _openAccounts = new List<AccountViewController>();
        Aura.Init("org.nickvision.money", "Nickvision Denaro");
        AppInfo.EnglishShortName = "Denaro";
        if (Directory.Exists($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}"))
        {
            // Move config files from older versions and delete old directory
            try
            {
                foreach (var file in Directory.GetFiles($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}"))
                {
                    File.Move(file, $"{UserDirectories.ApplicationConfig}{Path.DirectorySeparatorChar}{Path.GetFileName(file)}");
                }
            }
            catch (IOException) { }
            Directory.Delete($"{UserDirectories.Config}{Path.DirectorySeparatorChar}Nickvision{Path.DirectorySeparatorChar}{AppInfo.Name}", true);
        }
        AppInfo.Version = "2024.1.0";
        AppInfo.ShortName = _("Denaro");
        AppInfo.Description = _("Manage your personal finances");
        AppInfo.SourceRepo = new Uri("https://github.com/NickvisionApps/Denaro");
        AppInfo.IssueTracker = new Uri("https://github.com/NickvisionApps/Denaro/issues/new");
        AppInfo.SupportUrl = new Uri("https://github.com/NickvisionApps/Denaro/discussions");
        AppInfo.ExtraLinks[_("Matrix Chat")] = new Uri("https://matrix.to/#/#nickvision:matrix.org");
        AppInfo.Developers[_("Nicholas Logozzo")] = new Uri("https://github.com/nlogozzo");
        AppInfo.Developers[_("Contributors on GitHub ❤️")] = new Uri("https://github.com/NickvisionApps/Denaro/graphs/contributors");
        AppInfo.Designers[_("Nicholas Logozzo")] = new Uri("https://github.com/nlogozzo");
        AppInfo.Designers[_("Fyodor Sobolev")] = new Uri("https://github.com/fsobolev");
        AppInfo.Designers[_("DaPigGuy")] = new Uri("https://github.com/DaPigGuy");
        AppInfo.Artists[_("David Lapshin")] = new Uri("https://github.com/daudix-UFO");
        AppInfo.Artists[_("Tobias Bernard")] = new Uri("https://github.com/bertob");
        AppInfo.TranslatorCredits = _("translator-credits");
    }

    /// <summary>
    /// Finalizes the MainWindowController
    /// </summary>
    ~MainWindowController() => Dispose(false);

    /// <summary>
    /// Whether or not to show a sun icon on the home page
    /// </summary>
    public bool ShowSun
    {
        get
        {
            var timeNowHours = DateTime.Now.Hour;
            return timeNowHours >= 6 && timeNowHours < 18;
        }
    }

    /// <summary>
    /// The string for greeting on the home page
    /// </summary>
    public string Greeting
    {
        get
        {
            return DateTime.Now.Hour switch
            {
                >= 0 and < 6 => _p("Night", "Good Morning!"),
                < 12 => _p("Morning", "Good Morning!"),
                < 18 => _("Good Afternoon!"),
                < 24 => _("Good Evening!"),
                _ => _("Good Day!")
            };
        }
    }

    /// <summary>
    /// Whether or not to show graphs
    /// </summary>
    public bool ShowGraphs
    {
        get => Configuration.Current.ShowGraphs;

        set
        {
            Configuration.Current.ShowGraphs = value;
            Configuration.Current.Save();
        }
    }

    /// <summary>
    /// Frees resources used by the MainWindowController object
    /// </summary>
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Frees resources used by the MainWindowController object
    /// </summary>
    protected virtual void Dispose(bool disposing)
    {
        if (_disposed)
        {
            return;
        }
        if (disposing)
        {
            foreach (var controller in _openAccounts)
            {
                controller.Dispose();
            }
        }
        _disposed = true;
    }

    /// <summary>
    /// Creates a new PreferencesViewController
    /// </summary>
    /// <returns>The PreferencesViewController</returns>
    public PreferencesViewController CreatePreferencesViewController() => new PreferencesViewController();

    /// <summary>
    /// Creates a new NewAccountDialogController
    /// </summary>
    /// <returns>The PreferencesViewController</returns>
    public NewAccountDialogController CreateNewAccountDialogController() => new NewAccountDialogController(_openAccounts.Select(x => x.AccountPath));

    /// <summary>
    /// Creates a new DashboardViewController
    /// </summary>
    /// <returns>The DashboardViewController</returns>
    public DashboardViewController CreateDashboardViewController() => new DashboardViewController(_openAccounts, Configuration.Current.GroupDefaultColor);

    /// <summary>
    /// Gets an AccountViewController for the most recent account
    /// </summary>
    /// <returns>The AccountViewController</returns>
    public AccountViewController GetMostRecentAccountViewController() => _openAccounts[_openAccounts.Count - 1];

    /// <summary>
    /// Starts the application
    /// </summary>
    public async Task StartupAsync()
    {
        if (_fileToLaunch != null)
        {
            await AddAccountAsync(_fileToLaunch);
            _fileToLaunch = null;
        }
    }


    /// <summary>
    /// Creates an AccountViewController for the specified path
    /// </summary>
    /// <param name="path">The path of the open account</param>
    /// <returns>The AccountViewController or null if the account path is not open</returns>
    public AccountViewController? CreateAccountViewController(string path)
    {
        try
        {
            return _openAccounts[_openAccounts.FindIndex(x => x.AccountPath == path)];
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Gets a color for an account type
    /// </summary>
    /// <param name="accountType">The account type</param>
    /// <returns>The rgb color for the account type</returns>
    public string GetColorForAccountType(AccountType accountType)
    {
        return accountType switch
        {
            AccountType.Checking => Configuration.Current.AccountCheckingColor,
            AccountType.Savings => Configuration.Current.AccountSavingsColor,
            AccountType.Business => Configuration.Current.AccountBusinessColor,
            _ => Configuration.Current.AccountSavingsColor
        };
    }

    /// <summary>
    /// Gets whether or not an account with the given path is opened or not
    /// </summary>
    /// <param name="path">The path of the account to check</param>
    /// <returns>True if the account is open, else false</returns>
    public bool IsAccountOpen(string path) => _openAccounts.Any(x => x.AccountPath == path);

    /// <summary>
    /// Creates a new account and adds it to the list of opened accounts
    /// </summary>
    /// <param name="controller">NewAccountDialogController</param>
    /// <returns>True if new account created and opened, else false</returns>
    public async Task<bool> NewAccountAsync(NewAccountDialogController controller)
    {
        if (IsAccountOpen(controller.Path))
        {
            NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("Unable to overwrite an opened account."), NotificationSeverity.Error));
            return false;
        }
        if (File.Exists(controller.Path))
        {
            if (controller.OverwriteExisting)
            {
                File.Delete(controller.Path);
            }
            else
            {
                NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("Unable to overwrite an existing account."), NotificationSeverity.Error));
                return false;
            }
        }
        var accountViewController = new AccountViewController(controller.Path, NotificationSent, RecentAccountsChanged);
        accountViewController.TransferSent += OnTransferSent;
        try
        {
            accountViewController.Login(null);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("Unable to open the account. Please ensure that the app has permissions to access the file and try again."), NotificationSeverity.Error));
            return false;
        }
        _openAccounts.Add(accountViewController);
        AccountAdded?.Invoke(this, EventArgs.Empty);
        await Task.Delay(100);
        accountViewController.UpdateMetadata(controller.Metadata);
        if (!string.IsNullOrEmpty(controller.Password))
        {
            accountViewController.SetPassword(controller.Password, false);
        }
        if (File.Exists(controller.ImportFile))
        {
            await accountViewController.ImportFromFileAsync(controller.ImportFile);
        }
        return true;
    }

    /// <summary>
    /// Adds an account to the list of opened accounts
    /// </summary>
    /// <param name="path">The path of the account</param>
    /// <param name="showOpenedNotification">Whether or not to show a notification if an account is opened</param>
    /// <param name="password">A password for an account (if available)</param>
    /// <returns>True if account added, else false (account already added)</returns>
    public async Task<bool> AddAccountAsync(string path, bool showOpenedNotification = true, string? password = null)
    {
        if (Path.GetExtension(path).ToLower() != ".nmoney")
        {
            path += ".nmoney";
        }
        if (!_openAccounts.Any(x => x.AccountPath == path))
        {
            var controller = new AccountViewController(path, NotificationSent, RecentAccountsChanged);
            controller.TransferSent += OnTransferSent;
            try
            {
                if (controller.AccountNeedsPassword && string.IsNullOrEmpty(password))
                {
                    password = await AccountLoginAsync!(controller.AccountPath);
                }
                if (!controller.Login(password))
                {
                    controller.Dispose();
                    if (password != null)
                    {
                        NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("Unable to login to account. Provided password is invalid."), NotificationSeverity.Error));
                    }
                    return false;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
                NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("Unable to open the account. Please ensure that the app has permissions to access the file and try again."), NotificationSeverity.Error));
                return false;
            }
            _openAccounts.Add(controller);
            AccountAdded?.Invoke(this, EventArgs.Empty);
            return true;
        }
        else
        {
            if (showOpenedNotification)
            {
                NotificationSent?.Invoke(this, new NotificationSentEventArgs(_("This account is already opened."), NotificationSeverity.Warning));
            }
            return false;
        }
    }

    /// <summary>
    /// Closes the account with the provided index
    /// </summary>
    /// <param name="index">int</param>
    public void CloseAccount(int index)
    {
        _openAccounts[index].Dispose();
        _openAccounts.RemoveAt(index);
    }

    /// <summary>
    /// Removes a recent account from the list
    /// </summary>
    /// <param name="recentAccount">The RecentAccount to remove</param>
    public void RemoveRecentAccount(RecentAccount recentAccount)
    {
        Configuration.Current.RemoveRecentAccount(recentAccount);
        Configuration.Current.Save();
        RecentAccountsChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>
    /// Occurs when a transfer is sent from an account
    /// </summary>
    /// <param name="sender">object?</param>
    /// <param name="transfer">The transfer sent</param>
    private async void OnTransferSent(object? sender, Transfer transfer)
    {
        await AddAccountAsync(transfer.DestinationAccountPath, false, transfer.DestinationAccountPassword);
        var controller = _openAccounts.Find(x => x.AccountPath == transfer.DestinationAccountPath)!;
        while (!controller.IsOpened)
        {
            await Task.Delay(200);
        }
        await controller.ReceiveTransferAsync(transfer);
    }
}

File: libdenaro/src/controllers/mainwindowcontroller.cpp
﻿#include "controllers/mainwindowcontroller.h"
#include <ctime>
#include <locale>
#include <sstream>
#include <thread>
#include <libnick/app/aura.h>
#include <libnick/helpers/stringhelpers.h>
#include <libnick/localization/gettext.h>
#include "models/account.h"
#include "models/configuration.h"
#ifdef _WIN32
#include <windows.h>
#endif

using namespace Nickvision::Money::Shared::Models;
using namespace Nickvision::App;
using namespace Nickvision::Events;
using namespace Nickvision::Notifications;
using namespace Nickvision::Update;

namespace Nickvision::Money::Shared::Controllers
{
    MainWindowController::MainWindowController()
        : m_started{ false }
    {
#ifdef DEBUG
        Aura::getActive().init("org.nickvision.money", "Nickvision Denaro", "Denaro", Logging::LogLevel::Debug);
#else
        Aura::getActive().init("org.nickvision.money", "Nickvision Denaro", "Denaro", Logging::LogLevel::Info);
#endif
        AppInfo& appInfo{ Aura::getActive().getAppInfo() };
        appInfo.setVersion({ "2024.5.0-next" });
        appInfo.setShortName(_("Denaro"));
        appInfo.setDescription(_("Manage your personal finances"));
        appInfo.setSourceRepo("https://github.com/NickvisionApps/Denaro");
        appInfo.setIssueTracker("https://github.com/NickvisionApps/Denaro/issues/new");
        appInfo.setSupportUrl("https://github.com/NickvisionApps/Denaro/discussions");
        appInfo.getExtraLinks()[_("Matrix Chat")] = "https://matrix.to/#/#nickvision:matrix.org";
        appInfo.getDevelopers()["Nicholas Logozzo"] = "https://github.com/nlogozzo";
        appInfo.getDevelopers()[_("Contributors on GitHub")] = "https://github.com/NickvisionApps/Denaro/graphs/contributors";
        appInfo.getDesigners()["Nicholas Logozzo"] = "https://github.com/nlogozzo";
        appInfo.getDesigners()[_("Fyodor Sobolev")] = "https://github.com/fsobolev";
        appInfo.getDesigners()["DaPigGuy"] = "https://github.com/DaPigGuy";
        appInfo.getDesigners()["JoseBritto"] = "https://github.com/JoseBritto";
        appInfo.getArtists()[_("David Lapshin")] = "https://github.com/daudix";
        appInfo.getArtists()["Tobias Bernard"] = "https://github.com/bertob";
        appInfo.setTranslatorCredits(_("translator-credits"));
    }

    AppInfo& MainWindowController::getAppInfo() const
    {
        return Aura::getActive().getAppInfo();
    }

    bool MainWindowController::isDevVersion() const
    {
        return Aura::getActive().getAppInfo().getVersion().getVersionType() == VersionType::Preview;
    }

    Theme MainWindowController::getTheme() const
    {
        return Aura::getActive().getConfig<Configuration>("config").getTheme();
    }

    Event<EventArgs>& MainWindowController::configurationSaved()
    {
        return Aura::getActive().getConfig<Configuration>("config").saved();
    }

    Event<NotificationSentEventArgs>& MainWindowController::notificationSent()
    {
        return m_notificationSent;
    }

    Event<ShellNotificationSentEventArgs>& MainWindowController::shellNotificationSent()
    {
        return m_shellNotificationSent;
    }

    Event<ParamEventArgs<std::vector<RecentAccount>>>& MainWindowController::recentAccountsChanged()
    {
        return m_recentAccountsChanged;
    }

    Event<ParamEventArgs<const std::shared_ptr<AccountViewController>&>>& MainWindowController::accountAdded()
    {
        return m_accountAdded;
    }

    std::string MainWindowController::getDebugInformation(const std::string& extraInformation) const
    {
        std::stringstream builder;
        //App ID and Ver
        builder << Aura::getActive().getAppInfo().getId();
#ifdef _WIN32
        builder << ".winui" << std::endl;
#elif defined(__linux__)
        builder << ".gnome" << std::endl;
#endif
        builder << Aura::getActive().getAppInfo().getVersion().toString() << std::endl << std::endl;
        //System Information
        if(Aura::getActive().isRunningViaFlatpak())
        {
            builder << "Running under Flatpak" << std::endl;
        }
        else if(Aura::getActive().isRunningViaSnap())
        {
            builder << "Running under Snap" << std::endl;
        }
        else
        {
            builder << "Running locally" << std::endl;
        }
#ifdef _WIN32
        LCID lcid = GetThreadLocale();
        wchar_t name[LOCALE_NAME_MAX_LENGTH];
        if(LCIDToLocaleName(lcid, name, LOCALE_NAME_MAX_LENGTH, 0) > 0)
        {
            builder << StringHelpers::toString(name) << std::endl;
        }
#elif defined(__linux__)
        try
        {
            builder << std::locale("").name() << std::endl;
        }
        catch(...)
        {
            builder << "Locale not set" << std::endl;
        }
#endif
        //Gnuplot
        builder << std::endl << Aura::getActive().sysExec("gnuplot --version");
        //Extra Information
        if (!extraInformation.empty())
        {
            builder << std::endl << extraInformation << std::endl;
        }
        return builder.str();
    }

    std::string MainWindowController::getGreeting() const
    {
        std::time_t now{ std::time(nullptr) };
        std::tm* cal{ std::localtime(&now) };
        if (cal->tm_hour >= 0 && cal->tm_hour < 6)
        {
            return _p("Night", "Good Morning!");
        }
        else if (cal->tm_hour < 12)
        {
            return _p("Morning", "Good Morning!");
        }
        else if (cal->tm_hour < 18)
        {
            return _("Good Afternoon!");
        }
        else if (cal->tm_hour < 24)
        {
            return _("Good Evening!");
        }
        return _("Good Day!");
    }

    std::shared_ptr<PreferencesViewController> MainWindowController::createPreferencesViewController() const
    {
        return std::make_shared<PreferencesViewController>();
    }

    std::shared_ptr<NewAccountDialogController> MainWindowController::createNewAccountDialogController() const
    {
        return std::make_shared<NewAccountDialogController>();
    }

    std::shared_ptr<DashboardViewController> MainWindowController::createDashboardViewController() const
    {
        std::vector<std::shared_ptr<AccountViewController>> openAccounts;
        for(const std::pair<const std::filesystem::path, std::shared_ptr<AccountViewController>>& pair : m_accountViewControllers)
        {
            openAccounts.push_back(pair.second);
        }
        return std::make_shared<DashboardViewController>(openAccounts);
    }

    const std::shared_ptr<AccountViewController>& MainWindowController::getAccountViewController(const std::filesystem::path& path) const
    {
        return m_accountViewControllers.at(path);
    }

    void MainWindowController::startup()
    {
        if (!m_started)
        {
#ifdef _WIN32
            try
            {
                m_updater = std::make_shared<Updater>(Aura::getActive().getAppInfo().getSourceRepo());
            }
            catch(...)
            {
                m_updater = nullptr;
            }
            if (Aura::getActive().getConfig<Configuration>("config").getAutomaticallyCheckForUpdates())
            {
                checkForUpdates();
            }
#endif
            m_recentAccountsChanged.invoke({ Aura::getActive().getConfig<Configuration>("config").getRecentAccounts() });
            m_started = true;
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "MainWindow started.");
        }
    }

    void MainWindowController::checkForUpdates()
    {
        if(m_updater)
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Checking for updates...");
            std::thread worker{ [&]()
            {
                Version latest{ m_updater->fetchCurrentStableVersion() };
                if (!latest.empty())
                {
                    if (latest > Aura::getActive().getAppInfo().getVersion())
                    {
                        Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Update found: " + latest.toString());
                        m_notificationSent.invoke({ _("New update available"), NotificationSeverity::Success, "update" });
                    }
                    else
                    {
                        Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "No updates found.");
                    }
                }
                else
                {
                    Aura::getActive().getLogger().log(Logging::LogLevel::Warning, "Unable to fetch latest app version.");
                }
            } };
            worker.detach();
        }
    }

#ifdef _WIN32
    void MainWindowController::windowsUpdate()
    {
        if(m_updater)
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Fetching Windows app update...");
            std::thread worker{ [&]()
            {
                bool res{ m_updater->windowsUpdate(VersionType::Stable) };
                if (!res)
                {
                    Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Unbale to fetch Windows app update.");
                    m_notificationSent.invoke({ _("Unable to download and install update"), NotificationSeverity::Error, "error" });
                }
            } };
            worker.detach();
        }
    }

    void MainWindowController::connectTaskbar(HWND hwnd)
    {
        if(m_taskbar.connect(hwnd))
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Connected to Windows taskbar.");
        }
        else
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Unable to connect to Windows taskbar.");
        }
    }
#elif defined(__linux__)
    void MainWindowController::connectTaskbar(const std::string& desktopFile)
    {
        if(m_taskbar.connect(desktopFile))
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Connected to Linux taskbar.");
        }
        else
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Unable to connect to Linux taskbar.");
        }
    }
#endif

    bool MainWindowController::isAccountPasswordProtected(const std::filesystem::path& path) const
    {
        if(StringHelpers::toLower(path.extension().string()) != ".nmoney")
        {
            return false;
        }
        Account a{ path };
        return a.isEncrypted();
    }

    bool MainWindowController::hasOpenAccounts() const
    {
        return !m_accountViewControllers.empty();
    }

    void MainWindowController::newAccount(const std::shared_ptr<NewAccountDialogController>& newAccountDialogController)
    {
        if(std::filesystem::exists(newAccountDialogController->getFilePath()))
        {
            //Check if overwrite is allowed
            if(!newAccountDialogController->getOverwriteExisting())
            {
                Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Account exists and overwrite is disabled. (" + newAccountDialogController->getFilePath().string() + ")");
                m_notificationSent.invoke({ _("This account already exists."), NotificationSeverity::Error });
                return;
            }
            //Check if the account is open in the app (cannot delete if open)
            if(m_accountViewControllers.contains(newAccountDialogController->getFilePath()))
            {
                Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Cannot overwrite. Account is opened. (" + newAccountDialogController->getFilePath().string() + ")");
                m_notificationSent.invoke({ _("This account cannot be overwritten."), NotificationSeverity::Error });
                return;
            }
            Aura::getActive().getLogger().log(Logging::LogLevel::Warning, "Overwriting existing account file. (" + newAccountDialogController->getFilePath().string() + ")");
            std::filesystem::remove(newAccountDialogController->getFilePath());
        }
        Configuration& config{ Aura::getActive().getConfig<Configuration>("config") };
        //Create the new account
        std::unique_ptr<Account> account{ std::make_unique<Account>(newAccountDialogController->getFilePath()) };
        account->login("");
        account->setMetadata(newAccountDialogController->getMetadata());
        account->importFromFile(newAccountDialogController->getImportFile(), config.getTransactionDefaultColor(), config.getGroupDefaultColor());
        account->changePassword(newAccountDialogController->getPassword());
        account.reset();
        //Create the controller and open the account
        std::shared_ptr<AccountViewController> controller{ nullptr };
        try
        {
            controller = std::make_shared<AccountViewController>(newAccountDialogController->getFilePath(), newAccountDialogController->getPassword());
        }
        catch(const std::exception& e)
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Error, std::string(e.what()) + ". (" + newAccountDialogController->getFilePath().string() + ")");
            m_notificationSent.invoke({ e.what(), NotificationSeverity::Error });
        }
        if(controller)
        {
            m_accountViewControllers[newAccountDialogController->getFilePath()] = controller;
            config.addRecentAccount(controller->toRecentAccount());
            config.save();
            Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Config saved.");
            m_recentAccountsChanged.invoke({ config.getRecentAccounts() });
            m_accountAdded.invoke({ m_accountViewControllers[newAccountDialogController->getFilePath()] });
            Aura::getActive().getLogger().log(Logging::LogLevel::Info, "Account created. (" + newAccountDialogController->getFilePath().string() + ")");
        }
    }

    void MainWindowController::openAccount(std::filesystem::path path, const std::string& password)
    {
        //Check if the file is a Denaro account file
        if(StringHelpers::toLower(path.extension().string()) != ".nmoney")
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Error, "Invalid file extension. (" + path.string() + ")");
            m_notificationSent.invoke({ _("The file is not a Denaro account file."), NotificationSeverity::Error });
        }
        //Check if the account is already open
        else if(m_accountViewControllers.contains(path))
        {
            Aura::getActive().getLogger().log(Logging::LogLevel::Warning, "Account already open. (" + path.string() + ")");
            m_notificationSent.invoke({ _("The account is already open."), NotificationSeverity::Warning });
        }
        //Create the controller and open the account
        else
        {
            std::shared_ptr<AccountViewController> controller{ nullptr };
            try
            {
                controller = std::make_shared<AccountViewController>(path, password);
            }
            catch(const std::exception& e)
            {
                Aura::getActive().getLogger().log(Logging::LogLevel::Error, std::string(e.what()) + ". (" + path.string() + ")");
                m_notificationSent.invoke({ e.what(), NotificationSeverity::Error });
            }
            if(controller)
            {
                m_accountViewControllers[path] = controller;
                Configuration& config{ Aura::getActive().getConfig<Configuration>("config") };
                config.addRecentAccount(controller->toRecentAccount());
                config.save();
                Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Config saved.");
                m_recentAccountsChanged.invoke({ config.getRecentAccounts() });
                m_accountAdded.invoke({ m_accountViewControllers[path] });
                Aura::getActive().getLogger().log(Logging::LogLevel::Info, "Account opened. (" + path.string() + ")");
            }
        }
    }

    void MainWindowController::removeRecentAccount(const RecentAccount& account)
    {
        Configuration& config{ Aura::getActive().getConfig<Configuration>("config") };
        config.removeRecentAccount(account);
        config.save();
        Aura::getActive().getLogger().log(Logging::LogLevel::Debug, "Config saved.");
        m_recentAccountsChanged.invoke({ config.getRecentAccounts() });
    }
}
